### PR TITLE
Add breadcrumbs info to the categories tree

### PR DESCRIPTION
**Pull request by rmilecki**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | For user experience it's usually a good idea to treat (e.g. style) in some special way: <br>1. Currently viewed category<br>2. Parent categories (breadcrumbs)<br><br>It helps visitors to navigate through the site.<br><br>This change adds "breadcrumb" and "current" bool properties (set to true) to the relevant categories. This will allow to customze their treating with a front-end code.<br><br>Default template got updated too to add "current" class for relevant `<li>`. Custom templates (e.g. one from the "classic" theme) may also use "breadcrumb" property for unfolding relevant categories by default.<br><br>A similar solution for adding "current" CSS class is used in the _ps_mainmenu_.
| Type?         | new feature
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? |
| How to test?  | 1. Make sure you don't use a custom `ps_categorytree.tpl` (e.g. from the `classic` theme)<br>2. Navigate to some category<br>3. Using web browser developer tools verify that current category list item looks like: `<li class="current">`

### CODE REVIEW

**Comment by Hlavtox** [comment]
@rmilecki Good idea Rafael! Would you mind if we could improve this a bit?

If we want to make the current category visible, we also need to render the three “opened” if its somewhere deeper in the tree. So we need to know the full path to it. Back then, I usually followed these tutorials -
https://www.sunnytoo.com/8222/how-to-set-the-category-tree-links-module-to-expand-the-brunch-containing-current-category-in-prestashop-1-7 

Also, I think we should add a configuration switch to this module, something like “Keep current category open” or something like this.

And last question, what if we used $this->controller->getCategory() instead of the cookie value? That could be more “safe” way to get it maybe. We need this module to mark the current category in categorycontroller and productcontroller, and both have $category attribute we need. 👍

Ping @kpodemski for opinion

**Comment by rmilecki** [comment]
Thank you for your review @Hlavtox!

Code in that article you linked it rather hard to follow (no indents, copy & paste instead of simple diff, screenshots instead of text) but I see the point. I like that feature.

I see two possible implementations of such feature:

1. Generating array of categories "path". I'd rather use some simple recursive function for that, something like

```
    protected function getCurrentPath($category)
    {
        if (!method_exists($this->context->controller, 'getCategory') || !($current = $this->context->controller->getCategory())) {
            return null;
        }

        if ($category['id'] == $current->id) {
            return [ $category['id'] ];
        }

        foreach ($category['children'] as $subcategory) {
            $path = $this->getCurrentPath($subcategory);
            if ($path) {
                array_unshift($path, $category['id']);
                return $path;
            }
        }

        return null;
    }
```
```
Array
(
    [0] => 1
    [1] => 2
)
```

2. Adding new properties to the `categories` array. Maybe something like
```
Array
(
    [id] => 1
    [link] => https://foo.com/1-categories
    [name] => Categories
    [desc] => 
    [in_path] => true
    [current] => false
    [children] => Array
        (
            [2] => Array
                (
                    [id] => 2
                    [link] => https://foo.com/2-foo
                    [name] => Foo
                    [desc] => 
                    [in_path] => true
                    [current] => true
                    [children] => Array
                        (
                        )
                )
            [3] => Array
                (
                    [id] => 3
                    [link] => https://foo.com/3-bar
                    [name] => Bar
                    [desc] => 
                    [in_path] => false
                    [current] => false
                    [children] => Array
                        (
                        )
                )
        )
)
```

**Comment by SharakPL** [comment]
I agree with @Hlavtox . Relying on cookie isn't the best approach and we need to show whole active path to be able to open current category in the tree.

**Comment by Hlavtox** [comment]
@rmilecki Yeah that looks good! :-) `in_path` and `current` and we are golden. 

Maybe `in_path` would be enough, because `if (inpath and empty(children))` = our current category? Or better to be cleaner?

And don't forget to `Validate::isLoadedObject($current)` to check if there is not some nonsense in the `$current = $this->context->controller->getCategory()`

WDYT @SharakPL

**Comment by SharakPL** [comment]
> Maybe `in_path` would be enough, because `if (inpath and empty(children))` = our current category? Or better to be cleaner?
> 

Current category can have children

**Comment by Hlavtox** [comment]
@SharakPL You are right 👍 dumb me

**Comment by rmilecki** [comment]
Thank you guys for reviews!

I'll improve this pull request, let me just see first (by waiting few days) if #35 will end up accepted. That will save me solving some conflicts.

**Comment by SharakPL** [comment]
> 2. Adding new properties to the `categories` array. Maybe something like
> ```php
> Array
> (
>     [id] => 1
>     [link] => https://foo.com/1-categories
>     [name] => Categories
>     [desc] => 
>     [in_path] => true
>     [current] => false
>     [children] => Array
>         (
>             [2] => Array
>                 (
>                     [id] => 2
>                     [link] => https://foo.com/2-foo
>                     [name] => Foo
>                     [desc] => 
>                     [in_path] => true
>                     [current] => true
>                     [children] => Array
>                         (
>                         )
>                 )
>             [3] => Array
>                 (
>                     [id] => 3
>                     [link] => https://foo.com/3-bar
>                     [name] => Bar
>                     [desc] => 
>                     [in_path] => false
>                     [current] => false
>                     [children] => Array
>                         (
>                         )
>                 )
>         )
> )
> ```
> 

IMO that's the way to go.
Get whole tree with `current = false` & `inPath = false` for each category, then mark current category in a tree with `current = true` & `inPath = true` and all its parent with `inPath = true`. Then it would be just simple check in the template:
```smarty
{if $current} ...
{if $inPath} ...
```

without any complicated checks like the ones suggested on ST-Themes.

**Comment by rmilecki** [comment]
I updated this pull request. If needed I'll simply rebase #35 later.

**Comment by rmilecki** [comment]
@Hlavtox, @SharakPL: can you take a look at the updated pull request, please?

**Review by SharakPL — DISMISSED** [review]
LGTM

**Comment by Hlavtox** [comment]
@rmilecki Good job dude! 🚀

- I removed the cookie crap here - https://github.com/PrestaShop/ps_categorytree/pull/67, so there will always be a category, no need to check it. 👍
- Let's also wait for your other PR https://github.com/PrestaShop/ps_categorytree/pull/35.

Regarding this PR:
- What about adding a switch? So if the theme supports it, you can still disable it? Something like `BLOCK_CATEG_HIGHLIGHT_CURRENT`?
- Wouldn't the original name `in_path` be better? I find the `setBreadcrumbs` and `breadcrumbs` somehow confusing with the other breadcrumbs. 🤔

**Comment by rmilecki** [comment]
I've rebased my branch onto the latest `dev`. Thanks to changes from #35 and #68 this pull request looks much cleaner now. It is also independ of #67.

@Hlavtox: I'm not sure if we need something like `BLOCK_CATEG_HIGHLIGHT_CURRENT`. I think it should be up to the theme whether to include some extra `.current {}` styling. I don't think (m)any common UI parts get styled depending on config option value. I believe config options are mostly used to adjust behaviour.

I don't very strong opinion on `in_path` vs `breadcrumb`. I just thought that `in_path` may be not descriptive enough. At the same time I didn't want too long variable like `in_current_category_path`. Anyone else wants to comment on this? @SharakPL @kpodemski @matks?

**Comment by kpodemski** [comment]
cc @Hlavtox @SharakPL - you're more into this topic, I could use your feedback here

**Comment by Hlavtox** [comment]
@rmilecki Implemented on dev branch, closing this. Will be available in 3.0.0 :-)
